Add metrics builder

manototh opened the pull request:
Preview:
https://axiom-mano-metrics-builder.mintlify.app/query-data/metrics/overview
https://axiom-mano-metrics-builder.mintlify.app/query-data/metrics/query-metrics

tothmano committed:
Add info about support

tothmano committed:
First draft

tothmano committed:
Fix query metrics page

tothmano committed:
Get query language overview ready for review

tothmano committed:
Update overview.mdx

manototh committed:
Apply suggestions from code review

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

tothmano committed:
Hide unavailable features

tothmano committed:
Implement review

manototh committed:
Merge branch 'main' into mano/metrics

tothmano committed:
Merge branch 'main' into mano/metrics

tothmano committed:
Add dataset types, sample queries, migration

tothmano committed:
Add differences to migration

tothmano committed:
Add to overview

tothmano committed:
Finish overview page

tothmano committed:
Update datasets.mdx

manototh committed:
Update query-data/metrics/migrate-metrics.mdx

manototh committed:
Apply changes to align syntax

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

manototh committed:
Merge branch 'main' into mano/metrics

manototh committed:
Fix Vale

manototh committed:
Remove join

manototh committed:
Remove join 2

manototh committed:
Remove replace, comment out logical

manototh committed:
Remove running avg examples

manototh committed:
Remove currently unsupported features

manototh committed:
Correct syntax for align

manototh committed:
Update query-data/metrics/query-metrics.mdx

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

manototh committed:
Back to correct syntax

Update query-metrics.mdx

manototh committed:
Fix align syntax

manototh committed:
Add code highlighting

manototh committed:
Update reference/datasets.mdx

Co-authored-by: Arne Bahlo <arne@axiom.co>

manototh committed:
Add current limitations

manototh committed:
Delete space from allowed chars

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

manototh committed:
First draft

manototh committed:
Fixes

manototh committed:
Change to private preview

manototh committed:
Merge branch 'main' into mano/metrics-builder

Licenser commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+## Concepts
+
+- **Dataset:** A group of related metrics.
+- **Metric:** One-dimensional time series data.
```
```suggestion
- **Metric:** A collection of all series that measure the same metric.
```

my suggestion has horrible wording but I'm not sure how to better phrase it, but it's definitely not one dimensional metric has both the 'tags' dimension and the 'time' dimension; but I think mentioning dimensionality is going to be very confusing to users

manototh replied:
Thanks! Could you please let me know what you think? https://github.com/axiomhq/docs/pull/438/commits/bc1ab1c2ce824b615ed91e673288a3f7e35ae9aa

Licenser replied:
that's really great! I love it :D

Licenser commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+1. **Filter**: Applies conditions to a series by its tags.
+1. **Transformation** can be the following:
+    - **Mapping:** Maps the data to a new value.
+    - **Aggregation:** Aggregates the data over a given time to a single value.
```
```suggestion
    - **Aggregation:** Aggregates the data over a given time.
```

I would leave out 'to a single value' as it can be confusing as one can still have multiple series's and multiple values in time

Licenser commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+1. **Transformation** can be the following:
+    - **Mapping:** Maps the data to a new value.
+    - **Aggregation:** Aggregates the data over a given time to a single value.
+    - **Grouping:** Groups the data by a set of tags, combining overlapping values.
```
```suggestion
    - **Grouping:** Groups the data by a set of tags, combining values over different series that are measured at the same timestamp.
```

Licenser commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+
+1. Click the Query tab.
+1. Click **Builder** in the top left.
+1. In **Dataset**, select a dataset with OTel metrics data, and then select the metric you want to query.
```
```suggestion
1. In **Dataset**, select a OTel metrics dataset, and then select the metric you want to query.
```

very minor change but technically someonme could ingest json otel metrics into a events dataset and create a 'dataset with otel metrics data' that won't work here

Licenser commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+1. In **Dataset**, select a dataset with OTel metrics data, and then select the metric you want to query.
+1. To filter data:
+    1. Click **+** in the **Where** section.
+    1. Select the field where you want to filter for values.
```
```suggestion
    1. Select the tag where you want to filter for values.
```

Licenser reviewed (approved):
Just some really tiny nit pickings, I love it!

manototh committed:
Apply suggestions from code review

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

manototh committed:
Improve metric definition

benenharrington commented on query-data/metrics/overview.mdx:
```diff
@@ -0,0 +1,55 @@
+Support for OTel metrics is currently in private preview. For more information, see [Feature states](/platform-overview/roadmap#feature-states).
+</Info>
+
+<Prerequisites />
```
I realise this isn't being introduced by this PR, but it might be confusing that we're advising users that their token should have "permissions to update the dataset" in order to ingest data. We do have a 'Dataset' row in the permissions table with an 'Update' checkbox, but this covers update actions to the resource, for example changing the description. To ingest, you want to ensure the explicit ingest permission is granted.

manototh replied:
Thanks! Fixing in https://github.com/axiomhq/docs/pull/455

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+1. **Transformation** can be the following:
+    - **Mapping:** Maps the data to a new value.
+    - **Aggregation:** Aggregates the data over a given time.
+    - **Grouping:** Groups the data by a set of tags, combining values over different series that are measured at the same timestamp.
```
Would it make more sense to use the 'align' language here?

```suggestion
1. **Transformation**: Aligns the raw data with the chosen time step, and performs selected mappings and groupings:
    - **Mapping:** Maps the data to a new value.
    - **Align:** Aggregates the data to the chosen time step using the chosen aggregation.
    - **Grouping:** Groups the data by a set of tags using the chosen aggregation.
```

manototh replied:
@benenharrington Could you please check this? https://github.com/axiomhq/docs/pull/438/commits/e7c5858d1f354ccd3be799bf93312bf8ab299852

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+A typical metrics query contains the following:
+
+1. **Source**: Defines dataset and the metric to query.
+1. **Filter**: Applies conditions to a series by its tags.
```
```suggestion
1. **Filter**: Refine the set of series queried by matching against tag values.
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+
+1. Click the Query tab.
+1. Click **Builder** in the top left.
+1. In **Dataset**, select a OTel metrics dataset, and then select the metric you want to query.
```
```suggestion
1. In **Dataset**, select an OTel metrics dataset, and then select the metric you want to query.
```
I'm guessing there's a convention you're applying here, just commenting to learn!

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+<img src="/doc-assets/shots/example-metrics-query.png" alt="Example metrics query" />
+</Frame>
+
+This example queries the `axiom-dev.metrics` dataset’s `alertmanager_alerts` metric one hour before the current time. It filters results to events where `k8s.namespace.name` is `monitoring` and aggregates events over 30- second time windows into their average value.
```
```suggestion
This example queries the `axiom-dev.metrics` dataset’s `alertmanager_alerts` metric one hour before the current time. It filters results to events where `k8s.namespace.name` is `monitoring` and aggregates events over 30-second time windows into their average value.
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+
+## Elements of queries
+
+The following explains each element of an AxQL query.
```
I think we need to rework this section to remove references to AxQL for now. I'll add suggestions below that just delete the various code snippets as it still seems worthwhile to document the available operators, but open to alternatives.
```suggestion
The following explains each element of a metrics query.
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+```kusto
+<dataset>:<metric>
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+```kusto
+| where <tag> <operator> <value>
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+```kusto
+| where project == /.*metrics.*/ and code >= 200 and code < 300
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+// Set existing values to 0 if the value is < 0.4. Else set to 1.
+| map < 0.4
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+```kusto
+| align to <time_window> using <aggregation_function>
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+// Count the data points in the last hour
+| align to 1h using count
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+```kusto
+| group [by <tag1>, <tag2>] using <aggregation_function>
+```
+
````
```suggestion
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,192 @@
+| `sum`    | Sums values in each interval.         |
+| `min`    | Takes the minimum value per interval. |
+| `max`    | Takes the maximum value per interval. |
+| `count`  | Counts non-null values per interval.  |
```
```suggestion
| `avg`    | Averages values in each group.     |
| `sum`    | Sums values in each group.         |
| `min`    | Takes the minimum value per group. |
| `max`    | Takes the maximum value per group. |
| `count`  | Counts non-null values per group.  |
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
````diff
@@ -0,0 +1,192 @@
+
+// Group data by the `project` and `namespace` tags using the `sum` aggregation
+| group by project, namespace using sum
+```
````
```suggestion
```

manototh committed:
Add info about x-axiom-metrics-dataset header

manototh committed:
Apply suggestions from code review

Co-authored-by: benenharrington <42470628+benenharrington@users.noreply.github.com>

manototh committed:
Merge branch 'main' into mano/metrics-builder

manototh committed:
Implement review

manototh committed:
Add info about `application/x-protobuf` content type

manototh committed:
Rename sidenav of Traces

manototh committed:
Explain design choices better

manototh committed:
Add to intro, arch, and features pages

manototh committed:
Merge branch 'main' into mano/metrics-builder

manototh committed:
Add dashboard screenshot

Licenser commented on query-data/metrics/overview.mdx:
```diff
@@ -6,12 +6,20 @@ sidebarTitle: Overview
+
+MetricsDB is engineered from the ground up to embrace dimensional complexity:
+
+- **High-cardinality as a design principle**: Store metrics with high-cardinality dimensional tags: customer IDs, trace IDs, pod names, user agents, and other highly variable dimensions. Where other metrics databases penalize you with higher costs or degraded performance, MetricsDB treats high cardinality as a core capability.
```
Even if we don't suffer as much as other systems from high cardinality we should definitely not suggest storing `customer IDs` or `trace IDs` as tags, that is still an anti-pattern that will not make for a good experience either for us or the customers.

I would prefer not give any examples here as users tend to do strange things and giving an example implies we explicitly support that no matter how silly a user goes.

Licenser commented on query-data/metrics/overview.mdx:
```diff
@@ -38,18 +46,31 @@ You can use OTel metrics in dashboards and monitors the same way you use logs an
+
+- **Timestamp precision**: Truncate nanosecond timestamps to second precision. Second-level granularity is sufficient for metrics analysis (unlike traces which need finer resolution), and this optimization significantly improves compression ratios and query speed.
+- **Unified tag namespace**: Flatten resource, scope, and metric tags into a single namespace. This simplification makes queries more straightforward and enables faster dimensional filtering. You don’t need to remember which tags came from which scope.
+- **Unit normalization**: Convert the `unit` field to `otel.metric.unit` for consistent handling across all metric types.
```
"`unit` field" should be "`unit` attribute`"

Licenser commented on platform-overview/features.mdx:
```diff
@@ -16,6 +16,11 @@ mode: "wide"
 |  |  |  |  |
+| **MetricsDB** | - | - | Dedicated metrics database purpose-built for high-cardinality time-series data without the cost penalties of traditional metrics systems. |
+| **MetricsDB** | Ingest | [OpenTelemetry Metrics](/send-data/opentelemetry) | Native support for OpenTelemetry metrics protocol. |
+| **MetricsDB** | Storage | High-cardinality native | Store metrics with high-cardinality dimensional tags (customer IDs, trace IDs, container names) without performance degradation. |
```
```suggestion
| **MetricsDB** | Storage | High-cardinality native | Store metrics with high-cardinality dimensional tags without performance degradation. |
```

As mentioned elsewhere I'd avoid giving explicit examples and if we do definitely not customer and trace id's :)

Licenser commented on platform-overview/features.mdx:
```diff
@@ -16,6 +16,11 @@ mode: "wide"
 | **EventDB** | Query | [Virtual fields](/query-data/virtual-fields) | Ability to derive new values from data in real-time during queries without pre-structuring or transforming data during ingestion. |
 |  |  |  |  |
+| **MetricsDB** | - | - | Dedicated metrics database purpose-built for high-cardinality time-series data without the cost penalties of traditional metrics systems. |
+| **MetricsDB** | Ingest | [OpenTelemetry Metrics](/send-data/opentelemetry) | Native support for OpenTelemetry metrics protocol. |
```
```suggestion
| **MetricsDB** | Ingest | [OpenTelemetry Metrics](/send-data/opentelemetry) | Native support for OpenTelemetry (gRPC over HTTP) metrics protocol. |
```

manototh committed:
Implement review

manototh committed:
Merge branch 'main' into mano/metrics-builder

manototh committed:
Remove unsupported functions

manototh committed:
Apply suggestions from code review

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

manototh committed:
Implement reviews from Dom

https://www.notion.so/axiomhq/Reviewing-MetricsDB-29efc65a30198041a502f3eea1f5f346

manototh committed:
Implement review

manototh committed:
Add info about guarantees

benenharrington commented on introduction.mdx:
```diff
@@ -9,34 +9,42 @@ Trusted by 30,000+ organizations, from high-growth startups to global enterprise
 ## Components
 
-Axiom consists of two fundamental components:
+Axiom consists of three specialized data storage engines, each purpose-built for its specific workload:
```
This seems to imply console is a data storage engine? I think the language you want here is more along the lines of 'two purpose-built data stores supported by a unified console experience' (just an example to highlight the distinction)

benenharrington commented on platform-overview/features.mdx:
```diff
@@ -16,6 +16,11 @@ mode: "wide"
 |  |  |  |  |
+| **MetricsDB** | - | - | Dedicated metrics database purpose-built for high-cardinality time-series data without the cost penalties of traditional metrics systems. |
+| **MetricsDB** | Ingest | [OpenTelemetry Metrics](/send-data/opentelemetry) | Native support for OpenTelemetry (gRPC over HTTP) metrics protocol. |
+| **MetricsDB** | Storage | High-cardinality native | Store metrics with high-cardinality dimensional tags without performance degradation. |
```
```suggestion
| **MetricsDB** | Storage | High-cardinality native | Store metrics with high-cardinality tags without performance degradation. |
```

benenharrington commented on platform-overview/architecture.mdx:
```diff
@@ -8,92 +8,88 @@ description: "Technical deep-dive into Axiom’s distributed architecture."
+### MetricsDB storage
 
-**Query compilation**: The APL (Axiom Processing Language) query is parsed, optimized, and compiled into an execution plan. The compiler performs predicate pushdown, projection optimization, and identifies which blocks need to be read.
+MetricsDB uses a specialized columnar format engineered for time-series metrics with high-cardinality dimensions:
```
```suggestion
MetricsDB uses a specialized columnar format engineered for time-series metrics with high-cardinality tags:
```

benenharrington commented on query-data/metrics/overview.mdx:
```diff
@@ -0,0 +1,96 @@
+
+MetricsDB is engineered from the ground up to embrace dimensional complexity:
+
+- **High-cardinality as a design principle**: Store metrics with high-cardinality dimensional tags. Where other metrics databases penalize you with higher costs or degraded performance, MetricsDB treats high cardinality as a core capability.
```
```suggestion
- **High-cardinality as a design principle**: Store metrics with high-cardinality tags. Where other metrics databases penalize you with higher costs or degraded performance, MetricsDB treats high cardinality as a core capability.
```

benenharrington commented on query-data/metrics/overview.mdx:
```diff
@@ -0,0 +1,96 @@
+
+MetricsDB applies the following transformations to improve query performance and reduce storage costs:
+
+- **Timestamp precision**: Truncate nanosecond timestamps to second precision. Second-level granularity is sufficient for metrics analysis (unlike traces which need finer resolution), and this optimization significantly improves compression ratios and query speed.
```
```suggestion
- **Timestamp precision**: Truncate nanosecond timestamps to second precision. MetricsDB is built for use cases where second-level granularity is sufficient, and this optimization significantly improves compression ratios and query speed.
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,123 @@
+| `abs`                 | Returns the absolute value of each data point.                   |
+| `interpolate::linear` |  Linear interpolation of missing values. |
+| `fill::prev`          | Fills missing values using the previous non-null value.          |
+| `!`                   | Negates the value. |
```
```suggestion
| `!`                   | Negates the value. The value 0 is mapped to 1, and all other values are mapped to 0. |
```

benenharrington commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,123 @@
+
+For example:
+- To calculate the number of series, use `group using count`.
+- To calculate the total number of nodes, use `group using sum`.
```
```suggestion
To sum the values of all series, use `group using sum`.
```

manototh replied:
@benenharrington Could you please explain why you would remove this example?

benenharrington replied:
Resolved in slack

benenharrington commented on reference/datasets.mdx:
```diff
@@ -72,7 +72,13 @@ To create a dataset using the Axiom app, follow these steps:
+    - Select **Axiom Events** if you plan to use the dataset for event data that doesn’t follow OpenTelemetry (OTel) conventions.
+    - If you plan to send OTel data to the dataset, select the type of OTel data you want to use the dataset for. You must use dedicated datasets for each OTel component. For more information, see [Send OpenTelemetry data to Axiom](/send-data/opentelemetry).
+1. In **Data retention**, select for how long to store your data in this dataset. For more information, see [Specify data retention period](#specify-data-retention-period).
+1. In **Region**, select the region where you want to store your data. For more information, see [Regions](/reference/regions).
```
```suggestion
```
This likely won't be an option for initial metrics preview customers

manototh committed:
Apply suggestions from code review

Co-authored-by: benenharrington <42470628+benenharrington@users.noreply.github.com>

manototh committed:
Implement review

manototh committed:
Merge branch 'main' into mano/metrics-builder

manototh committed:
Implement review

manototh committed:
Update query-data/metrics/query-metrics.mdx

Co-authored-by: benenharrington <42470628+benenharrington@users.noreply.github.com>

manototh committed:
Change to public preview

manototh committed:
Change to public

Licenser commented on query-data/metrics/query-metrics.mdx:
```diff
@@ -0,0 +1,123 @@
+| `abs`                 | Returns the absolute value of each data point.                   |
+| `interpolate::linear` |  Linear interpolation of missing values. |
+| `fill::prev`          | Fills missing values using the previous non-null value.          |
+| `!`                   | Negates the value. Maps 0 to 1, and all other values to 0. |
```
```suggestion
```

As it was not used anywhere we decided to remove this from the preview to provide a better alternative later

manototh committed:
Update query-data/metrics/query-metrics.mdx

Co-authored-by: Heinz N. Gies <heinz@licenser.net>

manototh committed:
Fix supported metric types

manototh committed:
Merge branch 'main' into mano/metrics-builder

manototh committed:
Add sections about MetricsDB vs EventDB
